### ISSUE TITLE
schema sync issue in terminate_replace nemesis

### ISSUE THREAD

**Issue by amoskong** [scylladb/scylla#4647]
*Installation details*
Scylla version (or git commit hash): 3.1.0.rc2-0.20190630.acff367ea
Cluster size: 6
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-08f0f9bd6c5f677fb

Test job: http://jenkins.scylladb.com/job/scylla-3.1/job/longevity/job/longevity-50gb-4days/6/
test id: 179c22c3-f980-4bc2-a8bb-761e7254b19c （the living cluster can be found in eu-west-1 by searching with test id)

In terminate_and_replace_node nemesis, the new node failed to bootstrap, it keeps raising message of `storage_service - JOINING: waiting for schema information to complete`.

The new node was stuck there for 6 hours, until another new node is added by another new terminate_and_replace_node nemesis.

```
7-03 01:13:32| sdcm.nemesis.ChaosMonkey: >>>>>>>>>>>>>Started random_disrupt_method terminate_and_replace_node
7-03 01:13:32| sdcm.nemesis.ChaosMonkey: Set current_disruption -> TerminateAndReplaceNode Node longevity-tls-50gb-4d-3-1-db-node-179c22c3-2 [54.246.204.81 | 10.0.206.156] (seed: False)
7-03 01:13:32| sdcm.nemesis.ChaosMonkey: TerminateAndReplaceNode Node longevity-tls-50gb-4d-3-1-db-node-179c22c3-2 [54.246.204.81 | 10.0.206.156] (seed: False)


7-03 04:16:19| RemoteCmdRunner [centos@10.0.175.197]: Jul 03 04:16:19 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] storage_se
- JOINING: waomplete
7-03 04:16:19| RemoteCmdRunner [centos@10.0.105.70]: Jul 03 04:16:19 ip-10-0-105-70.eu-west-1.compute.internal scylla.bin[6747]:  [shard 7] compaction -
cting [/var/landard1-142a9c209d1d11e987b7000000000001/mc-1393-big-Data.db:level=0, /var/lib/scylla/data/keyspace_snappy/standard1-142a9c209d1d11e987b700
0001/mc-1351-
7-03 04:16:19| RemoteCmdRunner [centos@10.0.105.70]: Jul 03 04:16:19 ip-10-0-105-70.eu-west-1.compute.internal scylla.bin[6747]:  [shard 0] rpc - client
206.156:7001:t
7-03 04:16:19| RemoteCmdRunner [centos@10.0.226.255]: Jul 03 04:16:19 ip-10-0-226-255.eu-west-1.compute.internal scylla.bin[6646]:  [shard 0] rpc - clie
0.206.156:700ost
7-03 04:16:19| Wait for: None: timeout - 10800 seconds - expired
7-03 04:16:19| Cluster longevity-tls-50gb-4d-3-1-db-cluster-179c22c3 (AMI: ['ami-08f0f9bd6c5f677fb'] Type: i3.4xlarge): Node setup failed: Node longevit
y-tls-50gb-4d-3-1-db-node-179c22c3-8 [34.252.232.246 | 10.0.175.197] (seed: False)

from terminated db-node2:
Jul 03 01:01:39 ip-10-0-206-156.eu-west-1.compute.internal scylla.bin[6606]:  [shard 0] storage_service - DRAINING: starting drain process
...

Jul 03 01:16:27 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 03 01:16:27 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] rpc - client 10.0.206.156:7001: fail to connect: No route to host
....
storage_service - JOINING: waiting for schema information to complete
storage_service - JOINING: waiting for schema information to complete
storage_service - JOINING: waiting for schema information to complete
....


7-03 07:33:38,125 nemesis          L0544 INFO | sdcm.nemesis.ChaosMonkey: >>>>>>>>>>>>>Started random_disrupt_method terminate_and_replace_node

Jul 03 07:34:05 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] storage_service - Node 10.0.64.243 state jump to normal
Jul 03 07:34:05 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/system/peers-37f71aca7dc2383ba70672528af04d4f/mc-56-big-Data.db:level=0, /var/lib/scylla/data/system/peers-37f71aca7dc2383ba70672528af04d4f/system-peers-ka-42-Data.db:level=0, ]
...
storage_service - JOINING: Replacing a node with token(s): {...
token_metadata - Token 3937995184590873033 changing ownership from 10.0.206.156 to 10.0.175.197
```

**Comment by roydahan** [comment]
@amoskong please filter out the SCT behaviour from the issue, and report exactly the scylla issue.

**Comment by amoskong** [comment]
all db logs:
[db_log.tar.gz](https://github.com/scylladb/scylla/files/3355329/db_log.tar.gz)

longevity-tls-50gb-4d-3-1-db-node-179c22c3-1 [34.244.72.160 | 10.0.226.255] 
longevity-tls-50gb-4d-3-1-db-node-179c22c3-2 [54.246.204.81 | 10.0.206.156]  (terminated node)
longevity-tls-50gb-4d-3-1-db-node-179c22c3-3 [54.194.0.179 | 10.0.64.243] 
longevity-tls-50gb-4d-3-1-db-node-179c22c3-4 [34.247.81.11 | 10.0.57.177] 
longevity-tls-50gb-4d-3-1-db-node-179c22c3-5 [34.245.172.175 | 10.0.85.71] 
longevity-tls-50gb-4d-3-1-db-node-179c22c3-6 [52.17.164.175 | 10.0.105.70] 
longevity-tls-50gb-4d-3-1-db-node-179c22c3-7 [52.50.158.205 | 10.0.2.228] 
longevity-tls-50gb-4d-3-1-db-node-179c22c3-8 [34.252.232.246 | 10.0.175.197] try 1: New added node (stuck in waiting schema sync)
longevity-tls-50gb-4d-3-1-db-node-179c22c3-9 [63.32.59.120 | 10.0.250.129] try 2: another added node
longevity-tls-50gb-4d-3-1-db-node-179c22c3-10 [52.30.131.26 | 10.0.42.84]  try 3:  another added node
longevity-tls-50gb-4d-3-1-db-node-179c22c3-11 [52.30.243.34 | 10.0.13.243] The added node in another new terminate_replace nemesiss 

longevity-tls-50gb-4d-3-1-loader-node-179c22c3-1 [54.154.72.112 | 10.0.106.202] 
longevity-tls-50gb-4d-3-1-loader-node-179c22c3-2 [34.250.81.251 | 10.0.201.180] 
longevity-tls-50gb-4d-3-1-loader-node-179c22c3-3 [34.251.148.181 | 10.0.106.206] 
longevity-tls-50gb-4d-3-1-monitor-node-179c22c3-1 [34.240.203.237 | 10.0.254.138]

**Comment by amoskong** [comment]
The cluster had 6 nodes at the beginning:
```
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.0.64.243   34.22 GB   256          ?       c41be865-602e-4c3c-aac8-492ec1d7ced2  1a
UN  10.0.2.228    32.82 GB   256          ?       221ed914-2dec-4e08-87b7-8106bb018295  1a
UN  10.0.105.70   31.06 GB   256          ?       5eaa3799-17c8-4769-81f5-40d60dbca3d6  1a
UN  10.0.85.71    35.25 GB   256          ?       f3e68383-8871-4429-a308-037b89e8a209  1a
UN  10.0.206.156  37.06 GB   256          ?       f5b2cc9b-dd4d-411a-bb78-3d3ffff3b7e9  1a
UN  10.0.226.255  36.19 GB   256          ?       db469b3a-e1c3-4766-83b5-a3c62113b6d7  1a
```

10.0.206.156 node was removed from AWS, then a new node (10.0.175.197) was added to replace old node (10.0.206.156), but the new node failed to JOIN and to be up in 6 hours.

@tgrabiec 

test id: 179c22c3-f980-4bc2-a8bb-761e7254b19c （[the living cluster can be found in eu-west-1 by searching with test id](https://eu-west-1.console.aws.amazon.com/ec2/v2/home?region=eu-west-1#Instances:search=179c22c3-f980-4bc2-a8bb-761e7254b19c;sort=tag:Name))

**Comment by tgrabiec** [comment]
Events related to 10.0.64.243 on the bootstrapping node:
```
Jul 03 01:16:05 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] gossip - InetAddress 10.0.64.243 is now UP, status = NORMAL
Jul 03 01:16:15 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] migration_manager - Requesting schema pull from 10.0.64.243:0
Jul 03 01:16:15 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] migration_manager - Pulling schema from 10.0.64.243:0
Jul 03 01:16:15 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] migration_manager - Requesting schema pull from 10.0.64.243:0
Jul 03 01:16:15 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] migration_manager - Requesting schema pull from 10.0.64.243:0
Jul 03 01:16:15 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] migration_manager - Schema merge with 10.0.64.243:0 completed
Jul 03 01:16:15 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] migration_manager - Pulling schema from 10.0.64.243:0
Jul 03 01:16:15 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] migration_manager - Schema merge with 10.0.64.243:0 completed
Jul 03 07:34:05 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] storage_service - Node 10.0.64.243 state jump to normal
Jul 03 07:34:06 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] storage_service - Node 10.0.64.243 state jump to normal
Jul 03 07:34:11 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] rpc - client 10.0.64.243:7001: fail to connect: Connection refused
Jul 03 07:34:12 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] rpc - client 10.0.64.243:7001: fail to connect: Connection refused
Jul 03 07:34:14 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] rpc - client 10.0.64.243:7001: fail to connect: Connection refused
Jul 03 07:34:17 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] rpc - client 10.0.64.243:7001: fail to connect: Connection refused
Jul 03 07:34:29 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] gossip - InetAddress 10.0.64.243 is now DOWN, status = shutdown
```

The bootstrapping node unblocks when 10.0.64.243 goes down:
```
Jul 03 01:16:16 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] database - Schema version changed to 8aef0ffa-4c9f-3bac-ad52-e5d9847ed1df
...
JOINING: waiting for schema information to complete
Jul 03 07:34:29 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] gossip - InetAddress 10.0.64.243 is now DOWN, status = shutdown
Jul 03 07:34:29 ip-10-0-175-197.eu-west-1.compute.internal scylla.bin[6612]:  [shard 0] storage_service - JOINING: schema complete, ready to bootstrap
```

DOWN nodes are ignored for schema agreement so this suggests that the schema disagreement was due to 10.0.64.243. Checking the logs of 10.0.64.243, I can see that it had the same schema version as the bootstrapping node since the beginning of bootstrap:

```
Jul 03 00:49:23 ip-10-0-64-243.eu-west-1.compute.internal scylla.bin[25067]:  [shard 6] database - Schema version changed to 8aef0ffa-4c9f-3bac-ad52-e5d9847ed1df
```

One possible explanation is that due to a bug in the gossiper the bootstrapping node saw either NULL for SCHEMA_VERSION of 10.0.64.243 or it saw a value which came from the previous generation. The evidence is no longer available in the cluster.

It would help to invoke `nodetool gossipinfo` when the node is stuck.

**Comment by tgrabiec** [comment]
@amoskong What we could also do to get more insight is to run Scylla servers with debug level for `migration_manager` logger.

**Comment by amoskong** [comment]
I will reproduce the problem, enable migration_manager debug, and provide
the gossipinfo in stuck

**Comment by asias** [comment]
I saw similar issue with bootstrap a node node.
n1, n2 with a user keyspace myks2 with rf = 2.
When bootstrap n3, I saw:
```
storage_service - JOINING: waiting for schema information to complete
```
scylla commit: fca1ae69ff63f30809d4277dec5ae6d0f4c2ca5a

**Comment by asias** [comment]
With this patch:
```
ommit 2950179414f339a11e4222861ff70bab90bb480b (HEAD)
Author: Asias He <asias@scylladb.com>
Date:   Fri Jul 5 09:05:17 2019 +0800

    storage_service: Pull schema from live nodes in join_token_ring
    
    If the local node does not have schema agreement with other nodes, pull
    schema from all the live nodes.

diff --git a/service/storage_service.cc b/service/storage_service.cc
index cddc41a95..f8c775b3f 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -618,6 +618,9 @@ void storage_service::join_token_ring(int delay) {
         // (post CASSANDRA-1391 we don't expect this to be necessary very often, but it doesn't hurt to be careful)
         while (!get_local_migration_manager().have_schema_agreement()) {
             set_mode(mode::JOINING, "waiting for schema information to complete", true);
+            auto nodes = gossiper.get_live_members();
+            auto live_nodes = std::vector<gms::inet_address>(nodes.begin(), nodes.end());
+            service::get_local_migration_manager().sync_schema(_db.local(), live_nodes).get();
             sleep_abortable(std::chrono::seconds(1)).get();
         }
         set_mode(mode::JOINING, "schema complete, ready to bootstrap", true);

```
the issue is gone.

**Comment by amoskong** [comment]
I just reproduced this issue by job: http://jenkins.scylladb.com/job/scylla-staging/job/amos/job/longevity-50gb-4days/4/

Test id: b453a89d-bf5e-442a-8e23-c5fdbc2372ca
[Living cluster link](https://eu-west-1.console.aws.amazon.com/ec2/v2/home?region=eu-west-1#Instances:search=b453a89d-bf5e-442a-8e23-c5fdbc2372ca;sort=tag:Name)

```
[centos@ip-10-0-192-44 ~]$ nodetool gossipinfo
nodetool: Scylla API server HTTP GET to URL '/failure_detector/endpoints' failed: Not found
See 'nodetool help' or 'nodetool help <command>'.
[centos@ip-10-0-192-44 ~]$ systemctl status scylla-server
● scylla-server.service - Scylla Server
   Loaded: loaded (/usr/lib/systemd/system/scylla-server.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/scylla-server.service.d
           └─capabilities.conf, mounts.conf
   Active: active (running) since Fri 2019-07-05 19:28:50 UTC; 1h 32min ago
  Process: 6473 ExecStartPre=/usr/lib/scylla/scylla_prepare (code=exited, status=0/SUCCESS)
 Main PID: 6683 (scylla)
   Status: "starting storage service"
   CGroup: /system.slice/scylla-server.service
           └─6683 /usr/bin/scylla /opt/scylladb/bin/../libexec/scylla.bin --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --log-to-syslog 1 --log-to...

Jul 05 21:01:11 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:12 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:13 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:14 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:15 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:16 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:17 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:18 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:19 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
Jul 05 21:01:20 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
[centos@ip-10-0-192-44 ~]$ systemctl status scylla-jmx
● scylla-jmx.service - Scylla JMX
   Loaded: loaded (/usr/lib/systemd/system/scylla-jmx.service; disabled; vendor preset: disabled)
   Active: active (running) since Fri 2019-07-05 19:28:50 UTC; 1h 32min ago
 Main PID: 6715 (scylla-jmx)
   CGroup: /system.slice/scylla-jmx.service
           └─6715 /usr/lib/scylla/jmx/symlinks/scylla-jmx -Xmx256m -XX:+UseSerialGC -XX:+HeapDumpOnOutOfMemoryError -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.j...

Jul 05 19:28:50 ip-10-0-192-44.eu-west-1.compute.internal systemd[1]: Started Scylla JMX.
Jul 05 19:28:50 ip-10-0-192-44.eu-west-1.compute.internal scylla-jmx[6715]: Using config file: /etc/scylla/scylla.yaml
Jul 05 19:28:51 ip-10-0-192-44.eu-west-1.compute.internal scylla-jmx[6715]: Connecting to http://127.0.0.1:10000
Jul 05 19:28:51 ip-10-0-192-44.eu-west-1.compute.internal scylla-jmx[6715]: Starting the JMX server
Jul 05 19:28:51 ip-10-0-192-44.eu-west-1.compute.internal scylla-jmx[6715]: JMX is enabled to receive remote connections on port: 7199
```

**Comment by amoskong** [comment]
The new node keeps raise message:
```
Jul 05 21:16:33 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] migration_manager - Checking schema state for 10.0.213.243.
Jul 05 21:16:33 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] migration_manager - Checking schema state for 10.0.97.56.
Jul 05 21:16:33 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] migration_manager - Checking schema state for 10.0.100.25.
Jul 05 21:16:33 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] migration_manager - Checking schema state for 10.0.184.92.
Jul 05 21:16:33 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] migration_manager - Schema mismatch for 10.0.184.92 (0dd56d20-52c7-3efa-b834-481a6acc9dc4 != 5e57bd5a-e359-305d-8c80-086fca630294).
Jul 05 21:16:33 ip-10-0-192-44.eu-west-1.compute.internal scylla.bin[6683]:  [shard 0] storage_service - JOINING: waiting for schema information to complete
```

10.0.184.92 is the seed node.

```
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-1 [52.211.105.92 | 10.0.184.92]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-2 [34.244.71.240 | 10.0.100.25]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-3 [54.194.241.170 | 10.0.97.56]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-4 [52.213.39.115 | 10.0.84.183]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-5 [34.247.183.20 | 10.0.213.243]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-6 [54.171.185.102 | 10.0.70.241]
longevity-tls-50gb-4d-3-1-repl-loader-node-b453a89d-1 [54.229.38.114 | 10.0.207.251]
longevity-tls-50gb-4d-3-1-repl-loader-node-b453a89d-2 [52.31.33.2 | 10.0.253.231]
longevity-tls-50gb-4d-3-1-repl-loader-node-b453a89d-3 [34.246.222.31 | 10.0.50.164]
longevity-tls-50gb-4d-3-1-repl-monitor-node-b453a89d-1 [34.253.227.78 | 10.0.161.219]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-7 [34.241.189.223 | 10.0.198.16]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-8 [54.77.158.14 | 10.0.58.14]
longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-9 [34.244.207.234 | 10.0.192.44]
```

[nodetool-describecluster-on-existing-5-nodes.txt](https://github.com/scylladb/scylla/files/3364050/nodetool-describecluster-on-existing-5-nodes.txt)
[nodetool-gossipinfo-on-existing-5-nodes.txt](https://github.com/scylladb/scylla/files/3364051/nodetool-gossipinfo-on-existing-5-nodes.txt)

The schema version in `nodetool describecluster` are all same `0dd56d20-52c7-3efa-b834-481a6acc9dc4`

**Comment by amoskong** [comment]
[system_ks_of_seed_10-0-184-92.tar.gz](https://github.com/scylladb/scylla/files/3364076/system_ks_of_seed_10-0-184-92.tar.gz)
[scylla-ALL-data_hang_new_node.tar.gz](https://github.com/scylladb/scylla/files/3364077/scylla-ALL-data_hang_new_node.tar.gz)

**Comment by amoskong** [comment]
get coredump of hang node (new added node-9):

```
[centos@ip-10-0-192-44 ~]$ ps aux|grep scylla
scylla     6683 15.8  1.3 17180156336 1646316 ? Ssl  19:28  22:56 /usr/bin/scylla /opt/scylladb/bin/../libexec/scylla.bin --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 1-7,9-15
scylla     6715  0.0  0.1 3536864 187472 ?      Ssl  19:28   0:08 /usr/lib/scylla/jmx/symlinks/scylla-jmx -Xmx256m -XX:+UseSerialGC -XX:+HeapDumpOnOutOfMemoryError -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.host=localhost -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=7199 -Djava.rmi.server.hostname=localhost -Dcom.sun.management.jmxremote.rmi.port=7199 -Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder -jar /usr/lib/scylla/jmx/scylla-jmx-1.0.jar
centos    31750  0.0  0.0 112720  2320 pts/0    S+   21:53   0:00 grep --color=auto scylla
[centos@ip-10-0-192-44 ~]$ kill -SIGQUIT 6683
-bash: kill: (6683) - Operation not permitted
[centos@ip-10-0-192-44 ~]$ sudo kill -SIGQUIT 6683
[centos@ip-10-0-192-44 coredump]$ ls -lhh
total 538M
-rw-r-----. 1 root root 538M Jul  5 21:55 core.scylla.996.bcc1e723da98438994884cf8fd063d1b.6683.1562363660000000.gz
```

The coredump of hang node had been uploaded to 
https://console.cloud.google.com/storage/browser/upload.scylladb.com/scylla_issue_4647_7ca7b454-dc3e-4070-8fcf-3a5f14794ff0/?project=upload-179716&authuser=0
(private link)

**Comment by slivne** [comment]
@tgrabiec I assume this is a regression

**Comment by tgrabiec** [comment]
@slivne I'm not sure yet what's the root cause. If the same longevity with the same set of nemeses passed before, it would indicate a possible regression.

**Comment by amoskong** [comment]
I didn't reproduce this problem by repeated NodeTerminateAndReplace nemesis.
But the problem can be reproduced by normal longevity, which use
ChaosMonkey nemesis (calling nemesis randomly).
It occurred 100% by recently about 5 longevity-50gb-4days (random nemesis)

**Comment by slivne** [comment]
@tgrabiec is there any way to make progress ...

@amoskong please run the same longevity on 3.0.8 with the same tests and lets get a definitive answer this is a regression

**Comment by tgrabiec** [comment]
@slivne Yes.

The core dumps hold the answer, but they're useless due to https://github.com/scylladb/scylla/issues/4673

We should solve that. I am not familiar with packaging, so that's not me.

It looks like the bootstrapping node (and all other nodes as well) see that 10.0.184.92 has SCHEMA version which it had when it booted, but it changed the schema version later to match the one used by other nodes in the cluster. Other nodes didn't notice that. nodetool sees agreement because it doesn't query gossip but the database layer directly.

Looks like gossip entered a weird state on 10.0.184.92, or used a corrupted version when it broadcasted the initial schema version.

**Comment by tgrabiec** [comment]
@amoskong How can I get a log of all nemeses in this run?

**Comment by tgrabiec** [comment]
I see this in the logs of 10.0.184.92 after it restarts after SIGSEGV:

```
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-108553.log. 22282232 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-144115188075964431.log. 815096 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-54043195528554516.log. 1327096 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-72057594038036499.log. 14884856 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-126100789566482446.log. 21319672 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-18014398509590537.log. 12029944 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-108086391057000464.log. 24252408 bytes skipped.
 commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-180143985094928371.log. 22413304 bytes skipped.
 commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-198158383604410375.log. 18710520 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-162129586585446414.log. 15306744 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-36028797019072530.log. 19509240 bytes skipped.
 commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-234187180623374338.log. 29360120 bytes skipped.
commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-90071992547518482.log. 29229048 bytes skipped.
 commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-216172782113892367.log. 22544376 bytes skipped.
```

@elcallio is this expected after a crash?

**Comment by amoskong** [comment]
> @amoskong How can I get a log of all nemeses in this run?

We change job to use new pipeline and SCT had problem, it didn't archived the log anywhere when the job was abort.

/CC @bentsi @fruch 

----

Logs copied from the workspace (it might be cleaned by new job):
- nemesis log: https://storage.cloud.google.com/upload.scylladb.com/amos/scylla_issue_4647/nemesis.log?authuser=0
- nemesis & wait schema info message: https://storage.cloud.google.com/upload.scylladb.com/amos/scylla_issue_4647/nemesis-wait-schema-info.log?authuser=0
- db nodes log: https://storage.cloud.google.com/upload.scylladb.com/amos/scylla_issue_4647/db-nodes-log.tar.gz?authuser=0
- sct log file: https://storage.cloud.google.com/upload.scylladb.com/amos/scylla_issue_4647/sct-job.log.tar.gz?authuser=0

Nemesis info:
```
12:58:13,174 >>>Started random_disrupt_method nodetool_refresh
12:58:53,381 <<<Finished random_disrupt_method nodetool_refresh
13:07:58,520 >>>Started random_disrupt_method soft_reboot_node
13:11:15,293 <<<Finished random_disrupt_method soft_reboot_node
13:19:19,288 >>>Started random_disrupt_method modify_table
13:27:45,279 >>>Started random_disrupt_method destroy_data_then_repair
13:32:10,142 <<<Finished random_disrupt_method destroy_data_then_repair
13:40:44,313 >>>Started random_disrupt_method restart_with_resharding
14:03:59,239 <<<Finished random_disrupt_method restart_with_resharding
14:12:38,829 >>>Started random_disrupt_method restart_then_repair_node
14:30:55,806 <<<Finished random_disrupt_method restart_then_repair_node
14:39:45,226 >>>Started random_disrupt_method destroy_data_then_rebuild
14:53:59,779 <<<Finished random_disrupt_method destroy_data_then_rebuild
15:03:02,047 >>>Started random_disrupt_method destroy_data_then_rebuild
15:17:28,501 <<<Finished random_disrupt_method destroy_data_then_rebuild
15:26:12,156 >>>Started random_disrupt_method nodetool_drain
15:29:23,466 <<<Finished random_disrupt_method nodetool_drain
15:38:22,184 >>>Started random_disrupt_method terminate_and_replace_node
16:03:26,636 <<<Finished random_disrupt_method terminate_and_replace_node
16:12:26,821 >>>Started random_disrupt_method terminate_and_replace_node
16:49:39,307 <<<Finished random_disrupt_method terminate_and_replace_node
16:59:08,839 >>>Started random_disrupt_method restart_then_repair_node
17:40:10,989 <<<Finished random_disrupt_method restart_then_repair_node
17:50:21,770 >>>Started random_disrupt_method truncate
18:01:47,012 >>>Started random_disrupt_method modify_table
18:02:14,380 <<<Finished random_disrupt_method modify_table
18:11:56,625 >>>Started random_disrupt_method validate_hh_short_downtime
18:15:54,443 <<<Finished random_disrupt_method validate_hh_short_downtime
18:26:01,060 >>>Started random_disrupt_method nodetool_enospc
18:32:16,686 <<<Finished random_disrupt_method nodetool_enospc
18:42:09,041 >>>Started random_disrupt_method nodetool_decommission
<new node is stuck for waiting schema info>
```

**Comment by amoskong** [comment]
> I see this in the logs of 10.0.184.92 after it restarts after SIGSEGV:

Did you see the coredump occurred on node1 (10.0.184.92), which occurred around the stuck is started (07-05 19:25:35 UTC)?

The above coredump had been uploaded to https://console.cloud.google.com/storage/browser/upload.scylladb.com/core.scylla.996.55289e02c57c4572841d001538c38d31.6684.1562354735000000.gz/?project=upload-179716&authuser=0

```
$ less longevity-tls-50gb-4d-3-1-repl-db-node-b453a89d-1/coredump.log 
           PID: 6684 (scylla)
           UID: 996 (scylla)
           GID: 1001 (scylla)
        Signal: 11 (SEGV)
     Timestamp: Fri 2019-07-05 19:25:35 UTC (1min 49s ago)
  Command Line: /usr/bin/scylla /opt/scylladb/bin/../libexec/scylla.bin --blocked-reactor-notify-ms 500 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 1-7,9-15
    Executable: /opt/scylladb/libreloc/ld.so
 Control Group: /
       Boot ID: 55289e02c57c4572841d001538c38d31
    Machine ID: df877a200226bc47d06f26dae0736ec9
      Hostname: ip-10-0-184-92.eu-west-1.compute.internal
      Coredump: /var/lib/systemd/coredump/core.scylla.996.55289e02c57c4572841d001538c38d31.6684.1562354735000000
       Message: Process 6684 (scylla) of user 996 dumped core.
                
                Stack trace of thread 6695:
                #0  0x00000000024a32ba _ZN7seastar12shared_mutex4wakeEv (/opt/scylladb/libexec/scylla.bin)
```

--------
Btw I saw another core file in the node, the generate time of the file is `Jul  5 19:27`.

```
[centos@ip-10-0-184-92 ~]$ ls /var/lib/scylla/core.scylla.996.55289e02c57c4572841d001538c38d31.6684.1562354735000000  -lh
-rw-r-----. 1 centos root 112G Jul  5 19:27 /var/lib/scylla/core.scylla.996.55289e02c57c4572841d001538c38d31.6684.1562354735000000
```

> ```
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-108553.log. 22282232 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-144115188075964431.log. 815096 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-54043195528554516.log. 1327096 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-72057594038036499.log. 14884856 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-126100789566482446.log. 21319672 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-18014398509590537.log. 12029944 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-108086391057000464.log. 24252408 bytes skipped.
>  commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-180143985094928371.log. 22413304 bytes skipped.
>  commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-198158383604410375.log. 18710520 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-162129586585446414.log. 15306744 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-36028797019072530.log. 19509240 bytes skipped.
>  commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-234187180623374338.log. 29360120 bytes skipped.
> commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-90071992547518482.log. 29229048 bytes skipped.
>  commitlog_replayer - Corrupted file: /var/lib/scylla/commitlog/CommitLog-1-216172782113892367.log. 22544376 bytes skipped.
> ```
> 
> @elcallio is this expected after a crash?

**Comment by amoskong** [comment]
> @tgrabiec is there any way to make progress ...
> 
> @amoskong please run the same longevity on 3.0.8 with the same tests and lets get a definitive answer this is a regression

There is job running for 6 hours, so far everything is fine, I will keep updating you status/result.
- http://jenkins.scylladb.com/job/scylla-staging/job/amos/job/longevity-50gb-4days/6/console
- Monitor: http://34.242.124.8:3000/d/lVRdyoSWz/scylla-per-server-metrics-nemesis-3-0?refresh=30s&orgId=1&from=now-6h&to=now&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&editview=settings
- ami-037f9e1e861189609 (eu-west-1 Ireland)
- 3.0.8-0.20190628.23da53c4f (got from http://jenkins.scylladb.com/job/scylla-3.0/job/3.0-test/job/scylla-ami-test/118/)

**Comment by amoskong** [comment]
> > @tgrabiec is there any way to make progress ...
> > @amoskong please run the same longevity on 3.0.8 with the same tests and lets get a definitive answer this is a regression
> 
> There is job running for 6 hours, so far everything is fine, I will keep updating you status/result.

The job was aborted unexpectedly (avocado or builder memory issue), only few nemesises were executed. The issue didn't occur, total execute time: 10 hours.
Resubmitted a new job: http://jenkins.scylladb.com/job/scylla-staging/job/amos/job/longevity-50gb-4days/7/

> * http://jenkins.scylladb.com/job/scylla-staging/job/amos/job/longevity-50gb-4days/6/console
> * Monitor: http://34.242.124.8:3000/d/lVRdyoSWz/scylla-per-server-metrics-nemesis-3-0?refresh=30s&orgId=1&from=now-6h&to=now&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&editview=settings
> * ami-037f9e1e861189609 (eu-west-1 Ireland)
> * 3.0.8-0.20190628.23da53c4f (got from http://jenkins.scylladb.com/job/scylla-3.0/job/3.0-test/job/scylla-ami-test/118/)

**Comment by slivne** [comment]
@roydahan any luck with tomek's code ? I think you had luck without tomek's code (that doesn't help)

**Comment by roydahan** [comment]
I'm a little bit lost here.
@tgrabiec is this one on our side?
Are you waiting for something from @amoskong?

**Comment by tgrabiec** [comment]
@roydahan The analysis of the coredump was blocked by https://github.com/scylladb/scylla/issues/4673. Since that's fixed now, we can try to reproduce again (on a version which includes eb49fae02080f2b2e34fe4d771dba464be626fea) and take a core dump of the mismatched node.

The test should run with `trace` level for the `gossip` and `migration_manager` loggers.

**Comment by slivne** [comment]
@roydahan ping

**Comment by tgrabiec** [comment]
So, it turns out that #4673 is not solved after all. But running with trace-level logging on the `gossip` logger on each node and from the very beginning of boot may give new information.

**Comment by amoskong** [comment]
I had started a new job this morning.

Retest by latest 3.1 ami-08debfb69c4b0256b (us-east-1)
- it contains fix of coredump debug issue: https://github.com/scylladb/scylla/commit/eb49fae02080f2b2e34fe4d771dba464be626fea
- copy ami: ami-0327b39af152e59a7  (eu-west-1)
- cluster: e9d7463b-90d1-4af6-8ad3-e6ff7d809693
- http://jenkins.scylladb.com/job/scylla-staging/job/amos/job/longevity-50gb-4days/12/
- 3.1.0.rc2-0.20190715.eb49fae02

Current job status: Job is still running, not reproduced until now.

**Comment by amoskong** [comment]
There is a job running (enabled trace for migration_manager & gossip for all db nodes) , I will update you when it reproduced the problem.

**Comment by amoskong** [comment]
I still didn't reproduce this issue, after enabled trace for migration_manager & gossip for all db nodes), some nemesis (eg: drain) are very slow, there are other new issue occurred to break the job.

I will try to reproduce this issue by new job after addressed new issue.

**Comment by slivne** [comment]
@roydahan / @amoskong any update on this ?

**Comment by roydahan** [comment]
No, we stopped trying after several runs.
We added the check that Tomek suggested in the other issue to longevity but no longevity has run since.

**Comment by slivne** [comment]
closing this for now - if it does reproduce we can reopen

**Comment by tgrabiec** [comment]
It reproduced again on 3.1-rc4

**Comment by tzach** [comment]
@tgrabiec @roydahan I understand this is not a regression, pushing to 3.2
(please correct me if I'm wrong)

**Comment by tgrabiec** [comment]
@tzach Why do you think it's not a regression?

I think the root cause is https://github.com/scylladb/scylla/commit/c96ee9882ba50662b36b632349dfa051bea74990.

We call `update_schema_version()` after features are enabled and we recalculated the schema version. This method is not updating gossip though. The node will still use it's `database::version()` to decide on syncing, so it will not sync and stay inconsistent in gossip until the next schema change.

There is no actual schema inconsistency, but the joining node will think there is.

**Comment by tzach** [comment]
Thanks for the prompt correction. Keeping this on 3.1 then
